bundle/util.go

```diff
@@ -0,0 +1,396 @@
+	// parametersToSchema converts the apb parameters into a JSON Schema format.
+	createProperties, err := extractProperties(plan.Parameters)
+	if err != nil {
+		return Schema{}, err
+	}
+	createRequired := extractRequired(plan.Parameters)
+
+	bindProperties, err := extractProperties(plan.BindParameters)
+	if err != nil {
+		return Schema{}, err
+	}
+	bindRequired := extractRequired(plan.BindParameters)
+
+	updatableProperties, err := extractUpdatable(plan.Parameters)
+	if err != nil {
+		return Schema{}, err
+	}
+	updatableRequired := extractUpdatableRequired(createRequired, updatableProperties)
+
+	// builds a Schema object for the various methods.
+	s := Schema{
+		ServiceInstance: ServiceInstanceSchema{
+			Create: map[string]*schema.Schema{
+				"parameters": {
+					SchemaRef:  schema.SchemaURL,
+					Type:       []schema.PrimitiveType{schema.ObjectType},
+					Properties: createProperties,
+					Required:   createRequired,
+				},
+			},
+			Update: map[string]*schema.Schema{
+				"parameters": {
+					SchemaRef:  schema.SchemaURL,
+					Type:       []schema.PrimitiveType{schema.ObjectType},
+					Properties: updatableProperties,
+					Required:   updatableRequired,
+				},
+			},
+		},
+		ServiceBinding: ServiceBindingSchema{
+			Create: map[string]*schema.Schema{
+				"parameters": {
+					SchemaRef:  schema.SchemaURL,
+					Type:       []schema.PrimitiveType{schema.ObjectType},
+					Properties: bindProperties,
+					Required:   bindRequired,
+				},
+			},
+		},
+	}
+
+	return s, nil
+}
+
+func extractProperties(params []ParameterDescriptor) (map[string]*schema.Schema, error) {
+	properties := make(map[string]*schema.Schema)
+
+	for _, pd := range params {
+		k := pd.Name
+
+		t, err := getType(pd.Type)
+		if err != nil {
+			return properties, err
+		}
+
+		properties[k] = &schema.Schema{
+			Title:       pd.Title,
+			Description: pd.Description,
+			Default:     pd.Default,
+			Type:        t,
+		}
+
+		setStringValidators(pd, properties[k])
+		setNumberValidators(pd, properties[k])
+		setEnum(pd, properties[k])
+	}
+
+	return properties, nil
+}
+
+func setStringValidators(pd ParameterDescriptor, prop *schema.Schema) {
+	if prop.Type[0] != schema.StringType {
+		return
+	}
+
+	// we can NOT set values on the Schema object if we want to be
+	// omitempty. Setting maxlength to 0 is NOT the same as omitting it.
+	// 0 is a worthless value for DeprecatedMaxlength so we will not set it
+
+	// maxlength
+	if pd.DeprecatedMaxlength > 0 {
+		prop.MaxLength = schema.Integer{Val: pd.DeprecatedMaxlength, Initialized: true}
+	}
+
+	// max_length overrides maxlength
+	if pd.MaxLength > 0 {
+		prop.MaxLength = schema.Integer{Val: pd.MaxLength, Initialized: true}
+	}
+	// min_length
+	if pd.MinLength > 0 {
+		prop.MinLength = schema.Integer{Val: pd.MinLength, Initialized: true}
+	}
+
+	// do not set the regexp if it does not compile
+	if pd.Pattern != "" {
+		patternRegex, err := regexp.Compile(pd.Pattern)
+		prop.Pattern = patternRegex
```

Comment:
This appears to be setting the pattern without checking that it compiled successfully.